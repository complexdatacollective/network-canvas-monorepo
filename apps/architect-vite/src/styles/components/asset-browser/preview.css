.asset-browser-preview .window__main {
	overflow-y: hidden;
}

.asset-browser-preview .window__content {
	height: 100%;
	width: 100%;
}

.asset-browser-preview .window__content div,
.asset-browser-preview .window__content audio,
.asset-browser-preview .window__content video {
	width: 100%;
	height: 100%;
}

.asset-browser-preview--type-network .window__main {
	overflow-x: auto;
	overflow-y: auto;
}

.asset-browser-preview--type-network .window__content {
	width: auto;
	height: auto;
}
