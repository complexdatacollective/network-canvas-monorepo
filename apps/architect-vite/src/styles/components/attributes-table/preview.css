.attributes-table-preview {
	display: flex;
	flex-direction: row;

	.attributes-table-preview__name,
	.attributes-table-preview__value,
	.attributes-table-preview__error,
	.attributes-table-preview__delete {
		padding: 0 var(--space-5);
	}

	.attributes-table-preview__name {
		flex: 0 0 15rem;
		font-weight: var(--font-weight-semibold);
	}

	.attributes-table-preview__value,
	.attributes-table-preview__error {
		flex-grow: 1;
	}

	.attributes-table-preview__error {
		font-style: italic;
		color: var(--error);
	}

	.attributes-table-preview__delete {
		flex: 0 0 var(--space-10);
		cursor: pointer;
		display: flex;
		justify-content: center;
		align-items: center;

		.icon {
			width: var(--space-5);
			height: var(--space-5);

			.cls-1,
			.cls-2 {
				fill: var(--text-dark);
			}
		}
	}
}
