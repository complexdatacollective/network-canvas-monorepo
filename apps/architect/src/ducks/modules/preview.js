import { getFormValues } from 'redux-form';
import { getWorkingPath } from '@selectors/session';
import { getProtocol } from '@selectors/protocol';
import previewDriver from '@app/utils/previewDriver';

const getStageIndex = (protocol, stageMeta) => {
  if (stageMeta.id) {
    return protocol.stages.findIndex(({ id }) => id === stageMeta.id);
  }

  if (stageMeta.insertAtIndex !== undefined) {
    return stageMeta.insertAtIndex;
  }

  return protocol.stages.length;
};

const getDraftStages = (protocol, stageMeta, draftStage) => {
  const { stages } = protocol;

  if (stageMeta.id) {
    return stages.map((stage) => {
      if (stage.id === draftStage.id) { return draftStage; }

      return stage;
    });
  }

  const stageIndex = getStageIndex(protocol, stageMeta);

  return [
    ...protocol.stages.slice(0, stageIndex),
    draftStage,
    ...protocol.stages.slice(stageIndex),
  ];
};

const SET_ZOOM = 'PREVIEW/ZOOM';
const REFRESH_PREVIEW = 'PREVIEW/REFRESH_PREVIEW';
const PREVIEW_DRAFT = 'PREVIEW/PREVIEW_DRAFT';
const CLOSE_PREVIEW = 'PREVIEW/CLOSE_PREVIEW';
const CLEAR_PREVIEW = 'PREVIEW/CLEAR_PREVIEW';

const zoom = (zoomFactor) => ({
  type: SET_ZOOM,
  zoom: zoomFactor,
});

const refresh = () => ({
  type: REFRESH_PREVIEW,
});

const closePreview = () => (dispatch) => {
  dispatch({ type: CLOSE_PREVIEW });
  previewDriver.close();
};

const clearPreview = () => (dispatch) => {
  dispatch({
    type: CLEAR_PREVIEW,
  });

  previewDriver.clear();
};

const previewDraft = (draft, stageIndex) => (dispatch, getState) => {
  const state = getState();

  const workingPath = getWorkingPath(state);

  const draftProtocol = {
    ...draft,
    /**
       * This allows assets to work correctly in the Network Canvas preview.
       *
       * Network canvas uses relative paths for the assets:// protocol, whereas
       * Architect uses full paths. Since Network Canvas prepares urls as:
       * `assets://${protocolUID}/assets/${asset}` this allows us to load files
       * from the correct location.
       */
    uid: workingPath,
  };

  dispatch({
    type: PREVIEW_DRAFT,
    draft: draftProtocol,
    stageIndex,
  });

  previewDriver.preview(draftProtocol, stageIndex);
};

const previewStageFromForm = (stageMeta, formName) => (dispatch, getState) => {
  const state = getState();
  const protocol = getProtocol(state);

  const draftStage = getFormValues(formName)(state);
  const stageIndex = getStageIndex(protocol, stageMeta);
  const draftStages = getDraftStages(protocol, stageMeta, draftStage);
  const draftProtocol = { ...protocol, stages: draftStages };

  dispatch(previewDraft(draftProtocol, stageIndex));
};

const actionTypes = {
  PREVIEW_DRAFT,
  SET_ZOOM,
  REFRESH_PREVIEW,
  CLEAR_PREVIEW,
};

const actionCreators = {
  closePreview,
  clearPreview,
  previewDraft,
  previewStageFromForm,
  zoom,
  refresh,
};

export {
  actionTypes,
  actionCreators,
};
