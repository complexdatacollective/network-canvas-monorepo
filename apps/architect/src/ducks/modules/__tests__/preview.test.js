/* eslint-env jest */

import thunk from 'redux-thunk';
import configureStore from 'redux-mock-store';
import { actionCreators } from '../preview';
import testState from '../../../__tests__/testState.json';

jest.mock('../../../utils/previewDriver');

const mockStore = configureStore([thunk]);

describe('preview', () => {
  describe('previewStageFromForm()', () => {
    let store;

    beforeEach(() => {
      store = mockStore(testState);
    });

    it('dispatches previewDraft() for stage id', () => {
      actionCreators.previewDraft = jest.fn();

      store.dispatch(actionCreators.previewStageFromForm(
        { id: 'pip' },
        'draft-by-id',
      ));

      const actions = store.getActions();
      const draftStages = actions[0].draft.stages;
      const { stageIndex } = actions[0];

      expect(stageIndex).toBe(1);
      expect(draftStages).toMatchSnapshot();
    });

    it('dispatches previewDraft() for insert at index', () => {
      actionCreators.previewDraft = jest.fn();

      store.dispatch(actionCreators.previewStageFromForm(
        { insertAtIndex: 2 },
        'draft-insert-at-index',
      ));

      const actions = store.getActions();
      const draftStages = actions[0].draft.stages;
      const { stageIndex } = actions[0];

      expect(stageIndex).toBe(2);
      expect(draftStages).toMatchSnapshot();
    });

    it('dispatches previewDraft() for insert at index 0', () => {
      actionCreators.previewDraft = jest.fn();

      store.dispatch(actionCreators.previewStageFromForm(
        { insertAtIndex: 0 },
        'draft-insert-at-index',
      ));

      const actions = store.getActions();
      const draftStages = actions[0].draft.stages;
      const { stageIndex } = actions[0];

      expect(stageIndex).toBe(0);
      expect(draftStages).toMatchSnapshot();
    });

    it('dispatches previewDraft() for end of stages', () => {
      actionCreators.previewDraft = jest.fn();

      const state = store.getState();
      const expectedStageIndex = state.protocol.present.stages.length;

      store.dispatch(actionCreators.previewStageFromForm(
        {},
        'draft-insert-at-end',
      ));

      const actions = store.getActions();
      const draftStages = actions[0].draft.stages;
      const { stageIndex } = actions[0];

      expect(stageIndex).toBe(expectedStageIndex);
      expect(draftStages).toMatchSnapshot();
    });
  });
});
