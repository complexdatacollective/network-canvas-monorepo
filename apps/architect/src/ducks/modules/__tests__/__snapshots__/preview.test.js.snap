// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`preview previewStageFromForm() dispatches previewDraft() for end of stages 1`] = `
Array [
  Object {
    "form": Object {
      "fields": Array [
        Object {
          "prompt": "message",
          "variable": "alpha",
        },
        Object {
          "prompt": "message",
          "variable": "1234-1234-4",
        },
      ],
    },
    "id": "bazz",
    "label": "bazz",
    "prompts": Array [
      Object {
        "id": "promptId",
        "text": "not a prompt",
      },
    ],
    "subject": Object {
      "entity": "node",
      "type": "bar",
    },
    "type": "NameGenerator",
  },
  Object {
    "form": Object {
      "fields": Array [
        Object {
          "prompt": "message",
          "variable": "bravo",
        },
        Object {
          "prompt": "message",
          "variable": "1234-1234-5",
        },
      ],
    },
    "id": "pip",
    "label": "pip",
    "prompts": Array [
      Object {
        "id": "promptId",
        "text": "not a prompt",
      },
    ],
    "subject": Object {
      "entity": "node",
      "type": "bar",
    },
    "type": "NameGenerator",
  },
  Object {
    "id": "buzz",
    "label": "buzz",
    "prompts": Array [
      Object {
        "id": "fizz",
        "text": "not a prompt",
      },
    ],
    "subject": Object {
      "entity": "node",
      "type": "foo",
    },
    "type": "NameGenerator",
  },
  Object {
    "id": "foxtrot",
    "label": "foxtrot",
    "prompts": Array [
      Object {
        "id": "golf",
        "text": "not a prompt",
      },
      Object {
        "id": "hotel",
        "text": "not a prompt",
      },
    ],
    "subject": Object {
      "entity": "node",
      "type": "pop",
    },
    "type": "NameGenerator",
  },
  Object {
    "id": "alpha",
    "label": "alpha",
    "prompts": Array [
      Object {
        "edges": Object {
          "create": "1234-5",
          "display": Array [
            "1234-5",
            "delta",
            "echo",
          ],
        },
        "id": "bravo",
        "text": "not a prompt",
      },
    ],
    "subject": Object {
      "entity": "edge",
      "type": "1234-5",
    },
    "type": "Sociogram",
  },
  Object {
    "id": "mike",
    "label": "mike",
    "prompts": Array [
      Object {
        "id": "oscar",
        "text": "not a prompt",
      },
    ],
    "type": "Sociogram",
  },
  Object {
    "background": Object {
      "image": "1234-asset-5",
    },
    "dataSource": "1234-asset-3",
    "id": "assets",
    "items": Array [
      Object {
        "content": "1234-asset-4",
        "id": "itemId",
        "type": "asset",
      },
    ],
    "label": "assets",
    "panels": Array [
      Object {
        "dataSource": "1234-asset-2",
        "id": "panelId",
        "title": "",
      },
    ],
    "prompts": Array [
      Object {
        "id": "promptId",
        "text": "not a prompt",
      },
    ],
    "type": "Information",
  },
  Object {
    "fizz": "pop",
  },
]
`;

exports[`preview previewStageFromForm() dispatches previewDraft() for insert at index 0 1`] = `
Array [
  Object {
    "bazz": "buzz",
  },
  Object {
    "form": Object {
      "fields": Array [
        Object {
          "prompt": "message",
          "variable": "alpha",
        },
        Object {
          "prompt": "message",
          "variable": "1234-1234-4",
        },
      ],
    },
    "id": "bazz",
    "label": "bazz",
    "prompts": Array [
      Object {
        "id": "promptId",
        "text": "not a prompt",
      },
    ],
    "subject": Object {
      "entity": "node",
      "type": "bar",
    },
    "type": "NameGenerator",
  },
  Object {
    "form": Object {
      "fields": Array [
        Object {
          "prompt": "message",
          "variable": "bravo",
        },
        Object {
          "prompt": "message",
          "variable": "1234-1234-5",
        },
      ],
    },
    "id": "pip",
    "label": "pip",
    "prompts": Array [
      Object {
        "id": "promptId",
        "text": "not a prompt",
      },
    ],
    "subject": Object {
      "entity": "node",
      "type": "bar",
    },
    "type": "NameGenerator",
  },
  Object {
    "id": "buzz",
    "label": "buzz",
    "prompts": Array [
      Object {
        "id": "fizz",
        "text": "not a prompt",
      },
    ],
    "subject": Object {
      "entity": "node",
      "type": "foo",
    },
    "type": "NameGenerator",
  },
  Object {
    "id": "foxtrot",
    "label": "foxtrot",
    "prompts": Array [
      Object {
        "id": "golf",
        "text": "not a prompt",
      },
      Object {
        "id": "hotel",
        "text": "not a prompt",
      },
    ],
    "subject": Object {
      "entity": "node",
      "type": "pop",
    },
    "type": "NameGenerator",
  },
  Object {
    "id": "alpha",
    "label": "alpha",
    "prompts": Array [
      Object {
        "edges": Object {
          "create": "1234-5",
          "display": Array [
            "1234-5",
            "delta",
            "echo",
          ],
        },
        "id": "bravo",
        "text": "not a prompt",
      },
    ],
    "subject": Object {
      "entity": "edge",
      "type": "1234-5",
    },
    "type": "Sociogram",
  },
  Object {
    "id": "mike",
    "label": "mike",
    "prompts": Array [
      Object {
        "id": "oscar",
        "text": "not a prompt",
      },
    ],
    "type": "Sociogram",
  },
  Object {
    "background": Object {
      "image": "1234-asset-5",
    },
    "dataSource": "1234-asset-3",
    "id": "assets",
    "items": Array [
      Object {
        "content": "1234-asset-4",
        "id": "itemId",
        "type": "asset",
      },
    ],
    "label": "assets",
    "panels": Array [
      Object {
        "dataSource": "1234-asset-2",
        "id": "panelId",
        "title": "",
      },
    ],
    "prompts": Array [
      Object {
        "id": "promptId",
        "text": "not a prompt",
      },
    ],
    "type": "Information",
  },
]
`;

exports[`preview previewStageFromForm() dispatches previewDraft() for insert at index 1`] = `
Array [
  Object {
    "form": Object {
      "fields": Array [
        Object {
          "prompt": "message",
          "variable": "alpha",
        },
        Object {
          "prompt": "message",
          "variable": "1234-1234-4",
        },
      ],
    },
    "id": "bazz",
    "label": "bazz",
    "prompts": Array [
      Object {
        "id": "promptId",
        "text": "not a prompt",
      },
    ],
    "subject": Object {
      "entity": "node",
      "type": "bar",
    },
    "type": "NameGenerator",
  },
  Object {
    "form": Object {
      "fields": Array [
        Object {
          "prompt": "message",
          "variable": "bravo",
        },
        Object {
          "prompt": "message",
          "variable": "1234-1234-5",
        },
      ],
    },
    "id": "pip",
    "label": "pip",
    "prompts": Array [
      Object {
        "id": "promptId",
        "text": "not a prompt",
      },
    ],
    "subject": Object {
      "entity": "node",
      "type": "bar",
    },
    "type": "NameGenerator",
  },
  Object {
    "bazz": "buzz",
  },
  Object {
    "id": "buzz",
    "label": "buzz",
    "prompts": Array [
      Object {
        "id": "fizz",
        "text": "not a prompt",
      },
    ],
    "subject": Object {
      "entity": "node",
      "type": "foo",
    },
    "type": "NameGenerator",
  },
  Object {
    "id": "foxtrot",
    "label": "foxtrot",
    "prompts": Array [
      Object {
        "id": "golf",
        "text": "not a prompt",
      },
      Object {
        "id": "hotel",
        "text": "not a prompt",
      },
    ],
    "subject": Object {
      "entity": "node",
      "type": "pop",
    },
    "type": "NameGenerator",
  },
  Object {
    "id": "alpha",
    "label": "alpha",
    "prompts": Array [
      Object {
        "edges": Object {
          "create": "1234-5",
          "display": Array [
            "1234-5",
            "delta",
            "echo",
          ],
        },
        "id": "bravo",
        "text": "not a prompt",
      },
    ],
    "subject": Object {
      "entity": "edge",
      "type": "1234-5",
    },
    "type": "Sociogram",
  },
  Object {
    "id": "mike",
    "label": "mike",
    "prompts": Array [
      Object {
        "id": "oscar",
        "text": "not a prompt",
      },
    ],
    "type": "Sociogram",
  },
  Object {
    "background": Object {
      "image": "1234-asset-5",
    },
    "dataSource": "1234-asset-3",
    "id": "assets",
    "items": Array [
      Object {
        "content": "1234-asset-4",
        "id": "itemId",
        "type": "asset",
      },
    ],
    "label": "assets",
    "panels": Array [
      Object {
        "dataSource": "1234-asset-2",
        "id": "panelId",
        "title": "",
      },
    ],
    "prompts": Array [
      Object {
        "id": "promptId",
        "text": "not a prompt",
      },
    ],
    "type": "Information",
  },
]
`;

exports[`preview previewStageFromForm() dispatches previewDraft() for stage id 1`] = `
Array [
  Object {
    "form": Object {
      "fields": Array [
        Object {
          "prompt": "message",
          "variable": "alpha",
        },
        Object {
          "prompt": "message",
          "variable": "1234-1234-4",
        },
      ],
    },
    "id": "bazz",
    "label": "bazz",
    "prompts": Array [
      Object {
        "id": "promptId",
        "text": "not a prompt",
      },
    ],
    "subject": Object {
      "entity": "node",
      "type": "bar",
    },
    "type": "NameGenerator",
  },
  Object {
    "foo": "bar",
    "id": "pip",
  },
  Object {
    "id": "buzz",
    "label": "buzz",
    "prompts": Array [
      Object {
        "id": "fizz",
        "text": "not a prompt",
      },
    ],
    "subject": Object {
      "entity": "node",
      "type": "foo",
    },
    "type": "NameGenerator",
  },
  Object {
    "id": "foxtrot",
    "label": "foxtrot",
    "prompts": Array [
      Object {
        "id": "golf",
        "text": "not a prompt",
      },
      Object {
        "id": "hotel",
        "text": "not a prompt",
      },
    ],
    "subject": Object {
      "entity": "node",
      "type": "pop",
    },
    "type": "NameGenerator",
  },
  Object {
    "id": "alpha",
    "label": "alpha",
    "prompts": Array [
      Object {
        "edges": Object {
          "create": "1234-5",
          "display": Array [
            "1234-5",
            "delta",
            "echo",
          ],
        },
        "id": "bravo",
        "text": "not a prompt",
      },
    ],
    "subject": Object {
      "entity": "edge",
      "type": "1234-5",
    },
    "type": "Sociogram",
  },
  Object {
    "id": "mike",
    "label": "mike",
    "prompts": Array [
      Object {
        "id": "oscar",
        "text": "not a prompt",
      },
    ],
    "type": "Sociogram",
  },
  Object {
    "background": Object {
      "image": "1234-asset-5",
    },
    "dataSource": "1234-asset-3",
    "id": "assets",
    "items": Array [
      Object {
        "content": "1234-asset-4",
        "id": "itemId",
        "type": "asset",
      },
    ],
    "label": "assets",
    "panels": Array [
      Object {
        "dataSource": "1234-asset-2",
        "id": "panelId",
        "title": "",
      },
    ],
    "prompts": Array [
      Object {
        "id": "promptId",
        "text": "not a prompt",
      },
    ],
    "type": "Information",
  },
]
`;
